File: src/main/java/org/telran/lecture_07_dynamic_array/practice/Dynamic_Array.js
/** Динамическич массив в реализации класса */

class DynamicArray {
    constructor(data = []) {
        this.array = new Array(data.length || 1); // Внутренний массив, для хранения элементов
        for (let i =0; i < data.length; i++){
            this.array[i] = data[i];
        }
        this.count = data.length; // Количество занятых ячеек массива
        this.size = data.length || 1;  // Физический размер массива в памяти
    }

    growSize() {
        let tmp = new Array(this.size * 2); // Создаем новый в 2 раза больше
        // Копируем элементы из старого
        for (let i = 0; i < this.size; i++) {
            tmp[i] = this.array[i];
        }
        this.array = tmp;
        this.size *= 2;
    }

    // function add an element at the end of array
    add(data) {
        if (this.count >= this.size) { // Есть места не хватает
            this.growSize();
        }
        this.array[this.count] = data;
        this.count++;
    }

    // function remove last element
    remove() {
        // TODO-2[complete]: реализовать метод
        if (this.count === 0) {
            throw new Error("Empty array");
        }
        this.count--;
    }

    // function add an element at given index
    addAt(index, data) {
        // TODO-3[complete part]: убрать дублирование кода
        if (this.count < this.size) { // Есть места хватает
            for (let i = this.count - 1; i >= index; i--) {
                this.array[i + 1] = this.array[i]; // сдвигаем все элементы вправо от текущего индекса
            }
            this.array[index] = data;
            this.count++;
        } else { // Есть места не хватает
            this.growSize();
            this.array[index] = data;
            this.count++;
        }

    }

    // function add an element at given index
    removeAt(index) {
        if (this.count > 0) {
            for (let i = index; i < this.count - 1; i++) {
                this.array[i] = this.array[i + 1]; // сдвигаем все элементы влево до текущего индекса
            }
            this.array[this.count - 1] = 0;
            this.count--;
        }
    }

    getArray() {
        return this.array;
    }

    set(index, data) {
        if (index >= this.count || index < 0) {
            throw Error("Индекс за границей массива");
        }
        this.array[index] = data;
    }

    clean() {
        this.array = new Array(1);
        this.count = 0;
        this.size = 1;
    }

    toString() {
        return `[${this.array.slice(0, this.count).join(", ")}]`;
    }

    length() {
        // TODO-1[complete]: реализуйте метод
        return this.count;
    }
}

// test DynamicArray
// let da = new DynamicArray();
let da = new DynamicArray([2, 5, 7, 7, 15]);
da.add(10); // 0
da.add(-5); // 1
da.add(5);  // 2
// da.remove();
console.log(`${da}`)
// da.set(2, 8);
// console.log(`${da}`)
// da.clean()
// console.log(`${da}`)
// console.log(da.length())

File: src/main/java/org/telran/lecture_07_dynamic_array/practice/DynamicArray.java
package org.telran.lecture_07_dynamic_array.practice;

import java.util.Arrays;

public class DynamicArray {
    private int[] array; // внутренний массив для хранения элементов
    private int count; // физический размер - сколько элементов храним
    private int size; // логический размер - сколько памяти выделили

    public DynamicArray() {
        array = new int[1];
        count = 0;
        size = 1;
    }


    // function add an element at the end of array
    public void add(int data) {
        if (count >= size) { // Есть места не хватает
            int[] tmp = new int[size * 2]; // Создаем новый в 2 раза больше
            // Копируем элементы из старого
            for (int i = 0; i < size; i++) {
                tmp[i] = array[i];
            }
            array = tmp;
            size *= 2;
        }
        array[count] = data;
        count++;
    }

    // function remove last element
    public void remove() {
        // TODO-2: реализовать метод
    }

    // function add an element at given index
    public void addAt(int index, int data) {
        if (count < size) { // Есть места хватает
            // TODO-3: убрать дублирование кода
            for (int i = count - 1; i >= index; i--) {
                array[i + 1] = array[i]; // сдвигаем все элементы вправо от текущего индекса
            }

            array[index] = data;
            count++;

        } else { // Есть места не хватает
            int[] tmp = new int[size * 2]; // Создаем новый в 2 раза больше
            // Копируем элементы из старого
            for (int i = 0; i < size; i++) {
                tmp[i] = array[i];
            }
            array = tmp;
            size *= 2;

            for (int i = count - 1; i >= index; i--) {
                array[i + 1] = array[i]; // сдвигаем все элементы вправо от текущего индекса
            }

            array[index] = data;
            count++;
        }
    }

    // function shift all element of right side from given index in left
    public void removeAt(int index) {
        if (count > 0) {
            for (int i = index; i < count - 1; i++) {
                array[i] = array[i + 1]; // сдвигаем все элементы влево до текущего индекса
            }
            array[count - 1] = 0;
            count--;
        }
    }

    public int[] getArray() {
        int[] result = new int[count];
        for (int i = 0; i < count; i++) {
            result[i] = array[i];
        }
        return result;
    }

    public int length() {
        // TODO-1: реализуйте метод
        return 0;
    }

    public static void main(String[] args) {
        DynamicArray data = new DynamicArray();
        data.add(1);
        data.add(2);
        data.add(3);

        System.out.println(Arrays.toString(data.getArray()));
    }
}
